File: Java/ReadersWriterLockJava/src/Reader.java
import javax.swing.*;
import java.util.concurrent.locks.Lock;

class Reader implements Runnable {
    private final Lock readLock;
    private final String name;
    private final int readingTime;
    private final JTextArea txtBox;

    public Reader(String name, Lock readLock, int readingTime, JTextArea txtBox) {
        this.name = name;
        this.readLock = readLock;
        this.readingTime = readingTime;
        this.txtBox = txtBox;
    }

    public Reader(String name, Lock readLock, JTextArea txtBox) {
        this(name, readLock, 250, txtBox);
    }

    public void read() {
        SwingUtilities.invokeLater(() -> txtBox.append(name + " begin "));
        try {
            Thread.sleep(readingTime);
        } catch (InterruptedException e) {
            txtBox.append("\nInterruptedException when reading: " + e + "\n");
            Thread.currentThread().interrupt();
        }
        SwingUtilities.invokeLater(() -> txtBox.append("\n" + name + " finish after reading " + readingTime + "ms\n"));
    }

    @Override
    public void run() {
        readLock.lock();
        try {
            read();
        } finally {
            readLock.unlock();
        }
    }
}

File: ReadersWriterLock/Reader.cs
namespace ReadersWriterLock
{
    internal class Reader
    {
        private readonly ILock readLock;
        private readonly string name;
        private readonly int readingTime;
        private readonly TextBox txtBox;

        public Reader(string name, ILock readLock, int readingTime, TextBox txtBox)
        {
            this.name = name;
            this.readLock = readLock;
            this.readingTime = readingTime;
            this.txtBox = txtBox;
        }

        public Reader(string name, ILock readLock, TextBox txtBox) : this(name, readLock, 250, txtBox)
        {
        }

        public void Read()
        {
            txtBox.BeginInvoke(new Action(() => txtBox.AppendText($"{name} begin ")));
            Thread.Sleep(readingTime);
            txtBox.BeginInvoke(new Action(() => txtBox.AppendText($"\r\n{name} finish after reading {readingTime}ms \r\n")));
        }

        public void Run()
        {
            readLock.Lock();
            try
            {
                Read();
            }
            catch (ThreadInterruptedException e)
            {
                txtBox.Invoke(new Action(() => txtBox.AppendText($"InterruptedException when reading: {e} \r\n")));
                Thread.CurrentThread.Interrupt();
            }
            finally
            {
                readLock.Unlock();
            }
        }
    }
}
